dspinner/pin.go

```diff
@@ -0,0 +1,1069 @@
+	for i := range ents {
+		var p *pin
+		p, err = decodePin(path.Base(ents[i].Key), ents[i].Value)
+		if err != nil {
+			return nil, err
+		}
+		pins[i] = p
+	}
+	return pins, nil
+}
+
+// ImportFromIPLDPinner converts pins stored in mdag based storage to pins
+// stores in the datastore. Returns a dspinner loaded with the exported pins,
+// and a count of the pins imported.
+//
+// After pins are stored in datastore, the root pin key is deleted to unlink
+// the pin data in the DAGService.
+func ImportFromIPLDPinner(dstore ds.Datastore, dserv ipld.DAGService, internal ipld.DAGService) (ipfspinner.Pinner, int, error) {
+	ipldPinner, err := ipldpinner.LoadPinner(dstore, dserv, internal)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	p := New(dstore, dserv).(*pinner)
+
+	ctx, cancel := context.WithTimeout(context.TODO(), loadTimeout)
+	defer cancel()
+
+	// Save pinned CIDs as new pins in datastore.
+	rCids, _ := ipldPinner.RecursiveKeys(ctx)
+	for i := range rCids {
+		err = p.addPin(rCids[i], ipfspinner.Recursive, "")
+		if err != nil {
+			return nil, 0, err
+		}
+	}
+	dCids, _ := ipldPinner.DirectKeys(ctx)
+	for i := range dCids {
+		err = p.addPin(dCids[i], ipfspinner.Direct, "")
+		if err != nil {
+			return nil, 0, err
+		}
+	}
+
+	// Delete root mdag key from datastore to remove old pin storage.
+	if err = dstore.Delete(pinDatastoreKey); err != nil {
+		return nil, 0, fmt.Errorf("cannot delete old pin state: %v", err)
+	}
+	if err = dstore.Sync(pinDatastoreKey); err != nil {
+		return nil, 0, fmt.Errorf("cannot sync old pin state: %v", err)
+	}
+
+	return p, len(rCids) + len(dCids), nil
+}
+
+// ExportToIPLDPinner exports the pins stored in the datastore by dspinner, and
+// imports them into the given internal DAGService.  Returns an ipldpinner
+// loaded with the exported pins, and a count of the pins exported.
+//
+// After the pins are stored in the DAGService, the pins and their indexes are
+// removed.
+func ExportToIPLDPinner(dstore ds.Datastore, dserv ipld.DAGService, internal ipld.DAGService) (ipfspinner.Pinner, int, error) {
+	p := New(dstore, dserv).(*pinner)
+	pins, err := p.loadAllPins()
+	if err != nil {
+		return nil, 0, fmt.Errorf("cannot load pins: %v", err)
+	}
+
+	ipldPinner := ipldpinner.New(dstore, dserv, internal)
+
+	seen := cid.NewSet()
+	for _, pp := range pins {
+		if seen.Has(pp.cid) {
+			// multiple pins not support; can only keep one
+			continue
+		}
+		seen.Add(pp.cid)
+		ipldPinner.PinWithMode(pp.cid, pp.mode)
+	}
+
+	ctx := context.TODO()
+
+	// Save the ipldpinner pins
+	err = ipldPinner.Flush(ctx)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Remove the dspinner pins and indexes
+	for _, pp := range pins {
+		p.removePin(pp)
+	}
+	err = p.Flush(ctx)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return ipldPinner, seen.Len(), nil
+}
+
+// LoadPinner loads a pinner and its keysets from the given datastore
+func LoadPinner(dstore ds.Datastore, dserv ipld.DAGService) (ipfspinner.Pinner, error) {
```

Comment:
This was done to follow existing code/logic, to avoid breakage in any code that was using one or the other function.

In the previous logic, `LoadPinner` would fail if there was no root key saved.  I think that `New` is called during ipfs init. 
Now that we do not have a root key, we do not need two separate functions.  I think there should only be one `LoadPinner` function.

Reply:
Done.
Removed `New` in favor of only having `LoadPinner`